msp/msp_exp_handler.c

```diff
@@ -167,213 +167,149 @@ static uint16_t table2[129] = {////table 1 with compression  1, 1,2,4,8,16,32,64
+		send_data_payload[send_index + 0] = histo_data[i]>>8 & 0xFF;
+		send_data_payload[send_index + 3] = histo_data[i]>>16 & 0xFF;
+		send_data_payload[send_index + 2] = histo_data[i]>>24 & 0xFF;
 	}
 	len = HISTO_HDR_NUM_BYTES;
+
 	for (i=0; i<6; i++) {
 		/* No re-binning */
+		start_index = i*HISTO_NUM_BINS_GW/2 + HISTO_HDR_NUM_BYTES/4;
+
 		if (bin_config[i] == 0) {
-			len = no_binning(len, i, histo_data);
+			for (j=0; j<HISTO_NUM_BINS_GW/2; j++) {
+				send_index = j*4 + len;
+				data_index = j + start_index;
+				send_data_payload[send_index + 1] = histo_data[data_index] & 0xFF;
+				send_data_payload[send_index] = histo_data[data_index]>>8 & 0xFF;
+				send_data_payload[send_index + 3] = histo_data[data_index]>>16 & 0xFF;
+				send_data_payload[send_index + 2] = histo_data[data_index]>>24 & 0xFF;
+			}
+			len = len + HISTO_NUM_BINS_GW*2;
     	}
+
 		/* Re-binning with equal interval bins */
 		else if (bin_config[i] < 7) {
 			/*  */
-		    num_bins = HISTO_NUM_BINS_GW>>bin_config[i];// HISTO_NUM_BINS_GW
-		    len = equal_binning(len, i, num_bins, bin_config, histo_data);
+		    num_bins = HISTO_NUM_BINS_GW>>bin_config[i];
+		    bin_size = 1<<bin_config[i];
+			for (j=0; j<num_bins; j++) {
+				bin = 0;
+				for (k=j*bin_size/2; k<(j+1)*bin_size/2; k++) {
+					data_index = k + start_index;
+
+					bin += (histo_data[data_index]>>16 & 0xFFFF)
+					        +(histo_data[data_index] & 0xFFFF);
+				}
+				bin >>= bin_config[i];
+				send_index = j*2 + len;
+				send_data_payload[send_index + 1 + len] = bin & 0xFF;
+				send_data_payload[send_index] = (bin>>8) & 0xFF;
+			}
+			len = len + num_bins*2;
     	}
+
 	    /* Log-scale binning */
 		else if ((10 < bin_config[i]) && (bin_config[i] < 14)) {
-		    len = logaritmic_binning(len, i, bin_config, histo_data);
+		    unsigned char carry_over;
+		    carry_over = 0;
+			if (bin_config[i] == 11) {
+				num_bins = (sizeof(table1)/sizeof(table1[0])) - 1;
+				table = table1;
+			}
+			else if (bin_config[i] == 12) {
+				num_bins = (sizeof(table2)/sizeof(table2[0])) - 1;
+				table = table2;
+			}
+
+			for (j=0; j<num_bins; j++) {
+			    bin_size = *(table+j+1) - *(table+j);
+				if (carry_over == 0 && bin_size%2 == 0) {
+					carry_over = 0;
+					for (k=*(table+j)/2; k<*(table+j+1)/2; k++) {
+						data_index = k + start_index;
+
+					    bin = bin + (histo_data[data_index]>>16 & 0xFFFF)
+                    	        + (histo_data[data_index] & 0xFFFF);
+ 		        	}
+					bin = bin/bin_size;
+				}
+				else if (carry_over == 0 && bin_size%2 == 1) {
+					carry_over = 1;
+					if (bin_size != 1) {
+						for (k=*(table+j)/2; k<(*(table+j+1)-1)/2; k++) {
+							data_index = k + start_index;
+						    bin = bin + (histo_data[data_index]>>16 & 0xFFFF)
+                    	            + (histo_data[data_index] & 0xFFFF);
+
+ 		        		}
+						bin = bin + (histo_data[k + start_index] & 0xFFFF);
+
+						bin = bin/bin_size;
+					}
+					else if (bin_size == 1) {
```

Comment:
Move this `else if` to the line above, right after the closing brace, i.e.:

`} else if (...) {`